CNF-3595: uninstall SR-IOV Network Operator (#40634)
openshift/openshift-docs: networking/hardware_networks/uninstalling-sriov-operator.adoc

[id="uninstalling-sriov-operator"]
= Uninstalling the SR-IOV Network Operator
include::modules/common-attributes.adoc[]
:context: uninstalling-sr-iov-operator

toc::[]

To uninstall the SR-IOV Network Operator, you must first delete any running SR-IOV workloads. You must also delete the webhooks after uninstalling the operator.

.Prerequisites

* Access to an {product-title} cluster using an account with `cluster-admin` permissions.
* The SR-IOV Network Operator Operator must be installed.

.Procedure

> . Delete all SR-IOV custom resources (CRs).

Reviewer: Due to [BZ2039950](https://bugzilla.redhat.com/2039950) we might have an issue with removing the CRs.
If the operator is installed from a subscription, the finalizers on the SriovNetwork might not be removed, and the deletion will hang on the remaining finalizer in SriovNetwork. Should we add this to the description?

Author: What concretely would you add? What would be the steps/commands?

Author: @mmirecki gentle reminder :-)

Reviewer: Sorry for the delay. Below is a suggestion on the note:

Note: Due to BZ2039950 the sriov operator might not perform a full cleanup after being removed. Some SriovNetworks might be stuck  pending deletetion due to stray finalizers. To remove these SriovNetworks please execute the following for each of the remaining networks:
`oc patch SriovNetwork -n <namespace> <network name>  -p '{"metadata":{"finalizers":null}}' --type=merge`

Reviewer: An automated way to patch all networks would be:
`oc get SriovNetwork -A -o name | sed 's:.*/::' | xargs oc patch SriovNetwork -n openshift-sriov-network-operator  -p '{"metadata":{"finalizers":null}}' --type=merge`

Reviewer: An alternative way to cleanup, replacing both this and the webhook changes would be to scale down the deployment and let it cleanup up:
oc scale deployment --replicas=0 -n openshift-sriov-network-operator                   sriov-network-operator

[id="uninstalling-sriov-operator"]
= Uninstalling the SR-IOV Network Operator
include::modules/common-attributes.adoc[]
:context: uninstalling-sr-iov-operator

toc::[]

To uninstall the SR-IOV Network Operator, you must first delete any running SR-IOV workloads. You must also delete the webhooks after uninstalling the operator.

.Prerequisites

* You must have access to an {product-title} cluster using an account with `cluster-admin` permissions.
* You have the SR-IOV Network Operator installed.

.Procedure

> . Delete all SR-IOV custom resources (CRs).

Reviewer: Could we enumarate the resources to remove?

oc delete sriovnetwork  -n openshift-sriov-network-operator --all
oc delete sriovnetworknodepolicy  -n openshift-sriov-network-operator --all
oc delete sriovibnetwork  -n openshift-sriov-network-operator --all

Maybe also add a note:
NOTE: it is imporant to delete these resource before removing the SRIOV operator. Failing to remove them will result in stray configuration item being left in the cluster.

Failing to remove networks will cause stray network attachment-defintions to stay, so it's important to delete this before removing the operator.

Author: Good suggestion on enumerating the resources to remove. Thanks for sharing the commands!

Regarding the note, I think it is unnecessary. We are describing the removal step-by-step, customers have no reasons to believe they can skip any step and the impact that would cause to the cluster.

> . Remove the SR-IOV Network Operator from your cluster by following the documentation on xref:../../operators/admin/olm-deleting-operators-from-cluster.adoc#olm-deleting-operators-from-a-cluster[deleting Operators from a cluster].

Reviewer 2: "Remove the SR-IOV Network Operator from your cluster by following the documentation on deleting Operators from a cluster." -> "Follow the instructions on Deleting operators from a cluster to remove the SR-IOV Network Operator from your cluster."

. The SR-IOV resource definitions (CRDs) remain in the cluster after the SR-IOV Network Operator is uninstalled. Delete the remaining SR-IOV CRDs:
+
[source,terminal]
----
$ oc get crd -o name | grep sriov | xargs oc delete
----

. Delete the SR-IOV webhooks:
+
[source,terminal]
----
$ oc delete mutatingwebhookconfigurations network-resources-injector-config
$ oc delete MutatingWebhookConfiguration sriov-operator-webhook-config
$ oc delete ValidatingWebhookConfiguration sriov-operator-webhook-config
----

. Delete references to stray finalizers from the SriovNetworks:
+
[source,terminal]
----
$ oc get SriovNetwork -A -o name | sed 's:.*/::' | xargs oc patch SriovNetwork -n openshift-sriov-network-operator  -p '{"metadata":{"finalizers":null}}' --type=merge
----

. Delete the SR-IOV Network Operator namespace:
+
[source,terminal]
----
$ oc delete namespace openshift-sriov-network-operator
----

> 

Reviewer 2: Suggestion: is there a way to check this task has been successfully accomplished.

Reviewer 3: It's better to include a confirmation step, if possible - we're often asked for them, and their use is well-defined in the mod docs guidelines. Validating that the actions were successful is never docs bloat. The sample syntax for the verification section is in [this template](https://raw.githubusercontent.com/redhat-documentation/modular-docs/master/modular-docs-manual/files/TEMPLATE_PROCEDURE_doing-one-procedure.adoc).

> [id="uninstalling-sriov-operator"]

Reviewer 3: You need the content type tag:
```suggestion
:_content-type: ASSEMBLY
[id="uninstalling-sriov-operator"]
```

= Uninstalling the SR-IOV Network Operator
include::modules/common-attributes.adoc[]
:context: uninstalling-sr-iov-operator

toc::[]

> To uninstall the SR-IOV Network Operator, you must first delete any running SR-IOV workloads. You must also delete the webhooks after uninstalling the operator.

Reviewer 3: Maybe:
```suggestion
To uninstall the SR-IOV Network Operator, you must delete any running SR-IOV workloads, uninstall the Operator, and delete the webhooks that the Operator used.
```

> .Prerequisites

Reviewer 3: You need to convert this content into a module and include it in the assembly. (Make sure you apply the `:_content-type: PROCEDURE` tag according to the [contrib guide](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#content-type-attributes).

* You must have access to an {product-title} cluster using an account with `cluster-admin` permissions.
* You have the SR-IOV Network Operator installed.

.Procedure

. Delete all SR-IOV custom resources (CRs):
+

> [source,terminal]

Reviewer 3: You can have only one command per [code block](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output) - if you have multiple, the copy function doesn't work. Please split up the commands.

----
$ oc delete sriovnetwork -n openshift-sriov-network-operator --all
$ oc delete sriovnetworknodepolicy -n openshift-sriov-network-operator --all
$ oc delete sriovibnetwork -n openshift-sriov-network-operator --all
----

> . Follow the instructions on xref:../../operators/admin/olm-deleting-operators-from-cluster.adoc#olm-deleting-operators-from-a-cluster[Deleting operators from a cluster] to remove the SR-IOV Network Operator from your cluster.

Reviewer 3: You shouldn't link to steps in another context like this. You should include them like in the [assembly template](https://raw.githubusercontent.com/redhat-documentation/modular-docs/master/modular-docs-manual/files/TEMPLATE_ASSEMBLY_a-collection-of-modules.adoc).

Author: I wonder how the page would look like by adding the include in the middle of a procedure step. I'm thinking I'll have to split the procedure into multiple procedure modules, similar to https://docs.openshift.com/container-platform/4.9/serverless/install/removing-openshift-serverless.html. Please share a change suggestion if you have one.

Also, as a documentation user, I find it out of place to include the whole OLM deleting operators module in the SR-IOV uninstall page. The instructions will be generic, referencing the Jaeger Operator as an example.

Waiting for further clarification.

Author: It might be relevant to note that there are 5 occurrences of the same xref pattern in the openshift-docs repository:

```
[openshift-docs]$ git grep -l "xref:../../operators/admin/olm-deleting-operators" 
cicd/gitops/uninstalling-openshift-gitops.adoc
cicd/pipelines/uninstalling-pipelines.adoc
distr_tracing/distr_tracing_install/distr-tracing-removing.adoc
jaeger/jaeger_install/rhbjaeger-removing.adoc
operators/admin/olm-status.adoc
virt/backup_restore/virt-installing-configuring-oadp.adoc
```

> . The SR-IOV resource definitions (CRDs) remain in the cluster after the SR-IOV Network Operator is uninstalled. Delete the remaining SR-IOV CRDs:

Reviewer 3: You should move the action of the step first, maybe:
```suggestion
. Delete the SR-IOV custom resource definitions that remain in the cluster after the SR-IOV Network Operator is uninstalled:
```

+
[source,terminal]
----

> $ oc get crd -o name | grep sriov | xargs oc delete

Reviewer 4: Unfortunately we can not use this command in order to remove sriov crds. 
This command also can remove these resources:  **customresourcedefinition.apiextensions.k8s.io/sriovfecclusterconfigs.sriovfec.intel.com
customresourcedefinition.apiextensions.k8s.io/sriovfecnodeconfigs.sriovfec.intel.com**
Both these resources are part of Intel operator. 

May be we just can list resources that should be removed:
- customresourcedefinition.apiextensions.k8s.io/sriovibnetworks.sriovnetwork.openshift.io
- customresourcedefinition.apiextensions.k8s.io/sriovnetworknodepolicies.sriovnetwork.openshift.io
- customresourcedefinition.apiextensions.k8s.io/sriovnetworknodestates.sriovnetwork.openshift.io
- customresourcedefinition.apiextensions.k8s.io/sriovnetworkpoolconfigs.sriovnetwork.openshift.io
- customresourcedefinition.apiextensions.k8s.io/sriovnetworks.sriovnetwork.openshift.io
- customresourcedefinition.apiextensions.k8s.io/sriovoperatorconfigs.sriovnetwork.openshift.io

----

. Delete the SR-IOV webhooks:
+

> [source,terminal]

Reviewer 3: Same here, only one command per code block.

----
$ oc delete mutatingwebhookconfigurations network-resources-injector-config
$ oc delete MutatingWebhookConfiguration sriov-operator-webhook-config
$ oc delete ValidatingWebhookConfiguration sriov-operator-webhook-config
----

> . Delete references to stray finalizers from the SriovNetworks:

Reviewer 3: "Stray" isn't used anywhere else in the docs collection, and our style guide suggests avoiding unique terms when possible. I suggest:
```suggestion
. Delete references to unused finalizers from the SriovNetworks:
```

+
[source,terminal]
----

> $ oc get SriovNetwork -A -o name | sed 's:.*/::' | xargs oc patch SriovNetwork -n openshift-sriov-network-operator  -p '{"metadata":{"finalizers":null}}' --type=merge

Reviewer 4: This command'll not work without crds. I don't think if it's needed.
